core/src/main/java/org/mule/runtime/core/api/util/FileUtils.java

```diff
@@ -321,15 +323,72 @@ public static void unzip(File archive, File directory, boolean verify) throws IO
-      } else if (namePath.normalize().toString().startsWith("..")) {
-        // Not specified, but presents a security risk (allows overwriting external files)
-        throw new InvalidZipFileException("External paths are not allowed: " + namePath.toString());
+      verifyZipEntryPath((ZipEntry) entries.nextElement());
+    }
+  }
+
+  public static void verifyZipEntryPath(ZipEntry entry) throws InvalidZipFileException {
+    Path namePath = Paths.get(entry.getName());
+    if (namePath.getRoot() != null) {
+      // According to .ZIP File Format Specification (Section 4.4.17), the path can not be absolute
+      throw new InvalidZipFileException("Absolute paths are not allowed: " + namePath.toString());
+    } else if (namePath.normalize().toString().startsWith("..")) {
+      // Not specified, but presents a security risk (allows overwriting external files)
+      throw new InvalidZipFileException("External paths are not allowed: " + namePath.toString());
+    }
+  }
+
+  /**
+   * Unzip the specified archive to the given directory. Equivalent to invoking {@link #unzip(InputStream, File, boolean)} with a
+   * {@code true} value for the {@code verify} parameter
+   *
+   * @param archive   the contents of the archive to be unzipped
+   * @param directory the target directory
+   */
+  public static void unzip(InputStream archive, File directory) throws IOException {
+    unzip(archive, directory, true);
+  }
+
+  /**
+   * Unzip the specified {@code archive} to the given {@code directory}.
+   *
+   * @param archive   the contents of the archive to be unzipped
+   * @param directory the target directory
+   * @param verify    whether to verify all entries before extractions
+   */
+  public static void unzip(InputStream archive, File directory, boolean verify) throws IOException {
+    if (directory.exists()) {
+      if (!directory.isDirectory()) {
+        throw new IOException("Directory is not a directory: " + directory);
+      }
+    } else {
+      if (!directory.mkdirs()) {
+        throw new IOException("Could not create directory: " + directory);
+      }
+    }
+
+    try (ZipInputStream zis = new ZipInputStream(archive)) {
+      ZipEntry entry;
+      while ((entry = zis.getNextEntry()) != null) {
+        if (verify) {
+          verifyZipEntryPath(entry);
+        }
+
+        File f = FileUtils.newFile(directory, entry.getName());
+        if (entry.isDirectory()) {
+          if (!f.exists() && !f.mkdirs()) {
+            throw new IOException("Could not create directory: " + f);
+          }
+        } else {
+          File file = new File(directory, entry.getName());
+          if (!file.getParentFile().exists() && !file.getParentFile().mkdirs()) {
+            throw new IOException("Unable to create folders for zip entry: " + entry.getName());
+          }
+
+          OutputStream os = new BufferedOutputStream(new FileOutputStream(f));
```

Comment:
Is the new FileOutputStream(f) closed by the BufferedOutputStream or we should close it?
A test for this would be nice